File: python/test/t_DistFunc_normal.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott

ot.TESTPREAMBLE()


# pNormal
xMin = 0.1
xMax = 0.9
nX = 10
grid = [0.0] * nX
for iX in range(nX):
    x = xMin + (xMax - xMin) * iX / (nX - 1)
    grid[iX] = x
    print(
        f"pNormal({x:.12g})={ot.DistFunc.pNormal(x):.6g}, complementary={ot.DistFunc.pNormal(x, True):.6g}"
    )
print("pNormal(", grid, "=", ot.DistFunc.pNormal(grid))
# qNormal
qMin = 0.0
qMax = 1.0
nQ = 11
grid = [0.0] * nQ
for iQ in range(nQ):
    q = qMin + (qMax - qMin) * iQ / (nQ - 1)
    grid[iQ] = q
    print(
        f"qNormal({q:.12g})={ot.DistFunc.qNormal(q):.6g}, complementary={ot.DistFunc.qNormal(q, True):.6g}"
    )
print("qNormal(", grid, ")=", ot.DistFunc.qNormal(grid))
# rNormal
nR = 10
for iR in range(nR):
    print(f"rNormal()={ot.DistFunc.rNormal():.6g}")
# pNormal2D
print(
    f"pNormal2D(1.0, 1.0, 0.9)={ot.DistFunc.pNormal2D(1.0, 1.0, 0.9):.6g}, complementary={ot.DistFunc.pNormal2D(1.0, 1.0, 0.9, True):.6g}"
)
print(
    f"pNormal2D(-1.0, -1.0, 1.0, 1.0, 0.9)={ot.DistFunc.pNormal2D(-1.0, -1.0, 1.0, 1.0, 0.9):.6g}"
)
# pNormal3D rectangular
print(
    f"pNormal3D(-1.0, -1.0, -1.0, 1.0, 1.0, 1.0, 0.5, 0.2, 0.0)={ot.DistFunc.pNormal3D(-1.0, -1.0, -1.0, 1.0, 1.0, 1.0, 0.5, 0.2, 0.0):.6g}"
)

# pNormalND (Genz algorithm)
print("MVN d=1 exact")
L = ot.TriangularMatrix(1)
L[0, 0] = 1.0
p = ot.DistFunc.pNormalND([-1.0], [1.0], L, 100)
expected = ot.DistFunc.pNormal(1.0) - ot.DistFunc.pNormal(-1.0)
ott.assert_almost_equal(p, expected)

print("MVN d=2 independent")
L = ot.TriangularMatrix(2)
L[0, 0] = 1.0
L[1, 0] = 0.0
L[1, 1] = 1.0
p = ot.DistFunc.pNormalND([-1.0, -1.0], [1.0, 1.0], L, 10000)
p1 = ot.DistFunc.pNormal(1.0) - ot.DistFunc.pNormal(-1.0)
expected = p1 * p1
ott.assert_almost_equal(p, expected, 1e-2)

print("MVN d=2 correlated")
sigma = ot.CovarianceMatrix(2)
sigma[0, 0] = 1.0
sigma[1, 0] = 0.5
sigma[1, 1] = 2.0
p = ot.DistFunc.pNormalND([-1.0, -2.0], [1.0, 2.0], sigma, 10000)
ott.assert_almost_equal(p, 0.5, 2e-1)

print("MVN d=2 with mu")
sigma = ot.CovarianceMatrix(2)
sigma[0, 0] = 1.0
sigma[1, 0] = 0.3
sigma[1, 1] = 1.0
p = ot.DistFunc.pNormalND([-1.0, -1.0], [1.0, 1.0], [0.5, -0.5], sigma, 10000)
ott.assert_almost_equal(p, 0.3808, 2e-1)

print("Error: dimension mismatch")
L = ot.TriangularMatrix(2)
L[0, 0] = 1.0
L[1, 0] = 0.0
L[1, 1] = 1.0
with ott.assert_raises(RuntimeError):
    ot.DistFunc.pNormalND([-1.0], [1.0], L)

print("Error: a > b")
L = ot.TriangularMatrix(1)
L[0, 0] = 1.0
with ott.assert_raises(TypeError):
    ot.DistFunc.pNormalND([1.0], [-1.0], L)

print("Error: n = 0")
L = ot.TriangularMatrix(1)
L[0, 0] = 1.0
with ott.assert_raises(TypeError):
    ot.DistFunc.pNormalND([-1.0], [1.0], L, 0)

print("Error: L(i,i) <= 0")
L = ot.TriangularMatrix(1)
L[0, 0] = 0.0
with ott.assert_raises(TypeError):
    ot.DistFunc.pNormalND([-1.0], [1.0], L)

print("Edge: a == b")
L = ot.TriangularMatrix(1)
L[0, 0] = 1.0
p = ot.DistFunc.pNormalND([0.0], [0.0], L)
ott.assert_almost_equal(p, 0.0, 0.0, 1e-15)

print("Edge: wide interval")
L = ot.TriangularMatrix(2)
L[0, 0] = 1.0
L[1, 0] = 0.0
L[1, 1] = 1.0
p = ot.DistFunc.pNormalND([-10.0, -10.0], [10.0, 10.0], L)
ott.assert_almost_equal(p, 1.0, 1.0, 1e-10)

print("MVN d=3 independent")
L = ot.TriangularMatrix(3)
L[0, 0] = 1.0
L[1, 0] = 0.0
L[2, 0] = 0.0
L[1, 1] = 1.0
L[2, 1] = 0.0
L[2, 2] = 1.0
p = ot.DistFunc.pNormalND([-1.0, -1.0, -1.0], [1.0, 1.0, 1.0], L, 10000)
p1 = ot.DistFunc.pNormal(1.0) - ot.DistFunc.pNormal(-1.0)
expected = p1 * p1 * p1
ott.assert_almost_equal(p, expected, 1e-2)

File: lib/test/t_DistFunc_normal.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class DistFunc for standard methods
 *
 *  Copyright 2005-2022 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
  {
    // Normal related functions
    {
      // pNormal
      Scalar xMin = 0.1;
      Scalar xMax = 0.9;
      UnsignedInteger nX = 10;
      Point grid(nX);
      for (UnsignedInteger iX = 0; iX < nX; ++iX)
      {
        Scalar x = xMin + (xMax - xMin) * iX / (nX - 1);
        grid[iX] = x;
        fullprint << "pNormal(" << x << ")=" << DistFunc::pNormal(x) << ", complementary=" << DistFunc::pNormal(x, true) << std::endl;
      }
      fullprint << "pNormal(" << grid << ")=" << DistFunc::pNormal(grid) << std::endl;
    } // pNormal
    {
      // qNormal
      Scalar qMin = 0.1;
      Scalar qMax = 0.9;
      UnsignedInteger nQ = 10;
      Point grid(nQ);
      for (UnsignedInteger iQ = 0; iQ < nQ; ++iQ)
      {
        Scalar q = qMin + (qMax - qMin) * iQ / (nQ - 1);
        grid[iQ] = q;
        fullprint << "qNormal(" << q << ")=" << DistFunc::qNormal(q) << ", complementary=" << DistFunc::qNormal(q, true) << std::endl;
      }
      fullprint << "qNormal(" << grid << ")=" << DistFunc::qNormal(grid) << std::endl;
    } // qNormal
    {
      // rNormal
      UnsignedInteger nR = 10;
      for (UnsignedInteger iR = 0; iR < nR; ++iR)
      {
        fullprint << "rNormal()=" << DistFunc::rNormal() << std::endl;
      }
    } // rNormal
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}
